src/arch/riscv/pagetable_walker.cc

```diff
@@ -195,6 +223,41 @@ Walker::WalkerState::initState(ThreadContext * _tc,
     assert(satp.mode == AddrXlateMode::SV39);
 }
 
+std::pair<bool, Fault>
+Walker::WalkerState::tryCoalesce(ThreadContext *_tc,
+    BaseMMU::Translation *translation,
+    const RequestPtr &_req, BaseMMU::Mode _mode)
+{
+    SATP _satp = _tc->readMiscReg(MISCREG_SATP);
+    bool priv_match = (
+        (mode == _mode) && (satp == _satp) &&
+        (pmode == walker->tlb->getMemPriv(_tc, _mode)) &&
+        (status == _tc->readMiscReg(MISCREG_STATUS)));
+
+    bool addr_match = false;
+    Addr addr_match_num = req->getVaddr();
+    Addr pre_match_num = _req->getVaddr();
+
+    addr_match = ((pre_match_num >> PageShift) << PageShift) ==
+                 ((addr_match_num >> PageShift) << PageShift);
+
+    if (priv_match && addr_match) {
+        // coalesce
+        DPRINTF(PageTableWalker,
+        "Coalescing walk for %#lx(pc=%#lx) into %#lx(pc=%#lx)\n",
+        req->getVaddr(), _req->getPC(), req->getVaddr(), req->getPC());
+        // add to list of requestors
+        Fault new_fault = NoFault;
+        if (timingFault != NoFault) {
+            // recreate fault for this txn, we don't have pmp yet
+            // TODO: also consider pmp's addr fault
+            DPRINTF(PageTableWalker, "FAULT in tryCoalesce\n");
+        }
+        return std::make_pair(true, new_fault);
```

Comment:
I'm a bit confused with this code. By my understanding, the coalesced walk will *always* return `NoFault`, even when `timingFault` is not `NoFault`. Is this the intended behavior?

Reply:
If the coalesced walk returns `Fault`, the previous walk should have already returned a `Fault`, but maybe this info that comes from a later coalesced walk could also be returned. I will fix this to make sure that this duplicate `Fault` will not be used later.

Reply:
My understanding is that you're coalescing a request with other requests when the new request is enqueued, and you keep track of the tag of each request.
I think we can handle the Fault as follows when a walk returns,
- In the case of NoFault, we can return NoFault to all of the coalesced requests.
- In the case of Fault, we can return Fault to one of the coalesced requests, and let the rest of the requests to try walking again.